src/halmos/sevm.py

```diff
@@ -2461,77 +2326,73 @@ def finalize(ex: Exec):
+                    slot: Word = ex.st.pop()
+                    ex.st.push(self.sload(ex, ex.this, slot))
+
                 elif opcode == EVM.SSTORE:
-                    self.sstore(ex, ex.this, ex.st.pop(), ex.st.pop())
+                    slot: Word = ex.st.pop()
+                    value: Word = ex.st.pop()
+                    self.sstore(ex, ex.this, slot, value)
 
                 elif opcode == EVM.RETURNDATASIZE:
-                    ex.st.push(con(ex.returndatasize()))
+                    ex.st.push(ex.returndatasize())
+
                 elif opcode == EVM.RETURNDATACOPY:
                     loc: int = ex.st.mloc()
                     offset: int = int_of(ex.st.pop(), "symbolic RETURNDATACOPY offset")
-                    # size (in bytes)
                     size: int = int_of(ex.st.pop(), "symbolic RETURNDATACOPY size")
 
-                    # TODO: do we need to pass returndatasize here?
-                    wstore_partial(
-                        ex.st.memory,
-                        loc,
-                        offset,
-                        size,
-                        ex.returndata(),
-                        ex.returndatasize(),
-                    )
+                    if loc + size > MAX_MEMORY_SIZE:
+                        raise HalmosException("RETURNDATACOPY > MAX_MEMORY_SIZE")
+
+                    if size > 0:
+                        if offset + size > ex.returndatasize():
+                            raise OutOfBoundsRead("RETURNDATACOPY out of bounds")
+
+                        data = ex.returndata().slice(offset, offset + size)
+                        ex.st.memory.set_slice(loc, loc + size, data)
 
                 elif opcode == EVM.CALLDATACOPY:
                     loc: int = ex.st.mloc()
                     offset: int = int_of(ex.st.pop(), "symbolic CALLDATACOPY offset")
-                    # size (in bytes)
                     size: int = int_of(ex.st.pop(), "symbolic CALLDATACOPY size")
+
+                    if size > MAX_MEMORY_SIZE:
+                        raise HalmosException("CALLDATACOPY size > MAX_MEMORY_SIZE")
+
                     if size > 0:
-                        calldata = ex.message().data
-                        if calldata is None:
-                            f_calldatacopy = Function(
-                                "calldatacopy_" + str(size * 8),
-                                BitVecSort256,
-                                BitVecSorts[size * 8],
-                            )
-                            data = f_calldatacopy(offset)
-                            wstore(ex.st.memory, loc, size, data)
-                        else:
-                            data = padded_slice(calldata, offset, size, con(0, 8))
-                            wstore_bytes(ex.st.memory, loc, size, data)
+                        data: ByteVec = ex.calldata()[offset : offset + size]
+                        ex.st.memory[loc : loc + size] = data
```

Comment:
which one do you prefer?

while I was doing the refactoring, `slice()` and `set_slice()` were convenient to figure out the places I still needed to convert, which is why it's now a little inconsistent. 

Using `slice()` and `set_slice()` is more verbose, but it's also easier to go to the function definition

Reply:
i agree on the convenience with slice() and set_slice(). i also personally prefer that, but this is probably because i'm not python native. :) i will defer to you!